File: examples/cpp-examples/Triangulation.dox.cpp
#include <MRMesh/MRPointCloud.h>
#include <MRMesh/MRPointCloudTriangulation.h>
#include <MRMesh/MRUniformSampling.h>
#include <MRVoxels/MROffset.h>

int main()
{
    // Generate point cloud
    MR::PointCloud pc;
    pc.points.reserve( 10000 );
    for ( auto i = 0; i < 100; ++i )
    {
        const auto u = MR::PI2_F * float( i ) / ( 100.f - 1.f );
        for ( auto j = 0; j < 100; ++j )
        {
            const auto v = MR::PI_F * float( j ) / ( 100.f - 1.f );

            pc.points.emplace_back(
                std::cos( u ) * std::sin( v ),
                std::sin( u ) * std::sin( v ),
                std::cos( v )
            );
        }
    }
    // Remove duplicate points
    auto vs = MR::pointUniformSampling( pc, {
        .distance = 1e-3f,
    } );
    assert( vs );
    pc.validPoints = std::move( *vs );
    pc.invalidateCaches();

    // Triangulate it
    auto triangulated = MR::triangulatePointCloud( pc );
    assert( triangulated );

    // Fix possible issues
    auto mesh = MR::offsetMesh( *triangulated, 0.f, { {
        .voxelSize = MR::suggestVoxelSize( *triangulated, 5e+6f ),
    } } );
}

File: examples/js-examples/Triangulation.dox.js
import createMeshLib from '@meshinspector/meshlib';

const ml = await createMeshLib();

// generate a point cloud sampling a unit sphere
using pointCloud = new ml.PointCloud();
const resolution = 100;
for (let i = 0; i < resolution; ++i) {
  const u = 2 * Math.PI * i / (resolution - 1);
  for (let j = 0; j < resolution; ++j) {
    const v = Math.PI * j / (resolution - 1);
    pointCloud.addPoint({
      x: Math.cos(u) * Math.sin(v),
      y: Math.sin(u) * Math.sin(v),
      z: Math.cos(v),
    });
  }
}

// remove the points that are too close (e.g. the duplicated poles)
using samplingSettings = new ml.UniformSamplingSettings();
samplingSettings.distance = 1e-3;
using samples = ml.pointUniformSampling(pointCloud, samplingSettings);
pointCloud.validPoints = samples;
pointCloud.invalidateCaches();

// triangulate the sampled cloud
using triangulationParams = new ml.TriangulationParameters();
using triangulated = ml.triangulatePointCloud(pointCloud, triangulationParams);

// fix possible issues with a zero offset
using offsetParams = new ml.OffsetParameters();
offsetParams.voxelSize = ml.suggestVoxelSize(triangulated, 5e6);
using mesh = ml.offsetMesh(triangulated, 0, offsetParams);

ml.MeshSave.toAnySupportedFormat(mesh, 'result.stl');
